### DISCUSSION TITLE
Custom generics and creative (uhh) uses of type annotations

### DISCUSSION THREAD

**Discussion by matejcik** [microsoft/pyright#4766] [General]
I'm trying to make a DSL for parsing binary structs that would work like this:
```
class Example(Struct):
     x: uint8  # actual type is int
     y: uint8[4]  # actual type is list[int]
     z: uint8[x]  # actual type is list[int]
```
This takes a surprisingly small amount of insanity to make work at run-time, but so far evaded all my attempts at type-checking. Which is sad, because my whole reason to attempt this was to take advantage of `dataclass_transform`.

The trouble is with the `uint8[4]` declaration: traditionally, an argument in brackets should be a type parameter, not a number or more exotic things. This seems to be the overall rule. Mypy will straight out refuse to accept anything that is not a type. Pyright allows me to try things, such as adding a metaclass and declaring an interesting type signature on its `__getitem__` -- but overall it seems less that I'm formulating a way to express myself in a type system, and more that i'm searching for the right combination of bugs.

What's frustrating about this is that pyright seems perfectly capable of figuring out that `uint8[4]` evaluates to a perfectly fine _type_ suitable for `y`, but this has an ugly interplay with the fact that `uint8` is itself something like `type[Field[int]]`, and `Field[int]` already has a type parameter and doesn't want "another one". I have made some progress with specifying a bunch of types manually, like IntField and BytesField, and then making a metaclass that is itself generic (so that it _even accepts_ a parameter in brackets) and overriding its `__getitem__ `, and that allows me to declare an array of ints, but not array of array of ints (which under the hood runs the same code!) ... but. Yeah. Combination of bugs and all.

It would be nice if something along these lines were explicitly supported, like, lift the restriction on what a type annotation is allowed to be, so that I can declare a class with a `__getitem__` and it allows me to write my own "generic-like" behavior.

Comments, tips, ideas?

---

Other possible approaches:
```
class Traditional(Struct):
    x = uint8()
    y = array(uint8)
```
Advantage: no difficulty with the type-checker. Given that `uint8` and `array` are the appropriate kinds of (generic) descriptor types, the `int` type of `x` and `list[int]` type of `y` is inferred with no trouble whatsoever.
Disadvantage: no annotations = no `dataclass_transform`. my `__init__(**kwargs)` stays sadly type-unchecked.

```
class Boring(Struct):
    x: int = uint8()
    y: list[int] = array(uint8, 4)
```
Advantage: works, does everything that I need.
Drawback: boring and repetitive. Using `uint8` descriptor unambiguously indicates that `x` is of type `int`, dtto for `array`. Why should I, and every user, have to repeat ourselves just so that the type-checker is happy?

I would be perfectly fine if there was a way to make `Traditional` approach work together with `dataclass_transform` or a similar thing -- is there? ...but given that `dataclass_transform`'s whole thing is "copy what dataclasses do pretty much to the letter", I assume that's not gonna happen.

**Comment by erictraut** [comment]
I'm not sure how you're declaring the symbol `uint8`. You haven't provided enough details.

Regardless, the expression `4` is not a valid type argument for a generic class in the Python type system, so this won't work. 

You can certainly use methods like `__getitem__` however you'd like at runtime, but what you're attempting here will not be compatible with the standard static type system in Python. Static type checkers like pyright and mypy will not work with your code.

**Reply by matejcik** [reply]
> Static type checkers like pyright and mypy will not work with your code.

see, that's the thing: the following code _will happily type-check today_ with pyright:
```python
class uint8meta(type):
    def __getitem__(self, key: int) -> "type[array[int]]":
        ...

class uint8(metaclass=uint8meta):
    def __get__(self, obj, owner = None) -> int:
        ...

    def __set__(self, obj, value: int) -> None:
        ...

class array(Generic[T]):
    def __get__(self, obj, owner = None) -> list[T]:
        ...

    def __set__(self, obj, value: list[T]) -> None:
        ...

class Example:
    x: uint8
    y: uint8[3]

e = Example()
reveal_type(e.x)  # int
reveal_type(e.y)  # list[int]
```
Of course, my guess is that now that i've told you, you will quickly identify the bug that caused this not to be rejected and fix it in some upcoming pyright release. But the capability is undoubtedly there. The type signatures say a true thing: indexing the `uint8` type will result in a different type, also a descriptor type, and the correct type is inferred at the end.

This is, at least on the surface, not too different from what generics are already doing. Indexing a type X with type T will evaluate to a new type called X[T]. :woman_shrugging: The limitation that only brackets are allowed in type annotations, and only for types, seems arbitrary.

The problems start when I try to say that `uint8` is actually itself a `Generic[T]`, because at that point my override of `__getitem__` is in conflict with the version coming from `Generic`. To "say a true thing" (now ignoring the actual constraints the typechecker presents), my `__getitem__` would need to be declared on, say `FieldType[T]`, but not on `FieldType`, i.e., I would need `FieldType[T]` to actually be its own type separate from the generic declaration.
Again, this is doable at runtime and it also seems expressible in the type annotation language. But for it to work, I would need to be able to chain indexing, i.e., `FieldType[T][3]` would need to be a valid expression in the annotation.

Looking at it more generally, one inroad for that could be some way of opting into creating my own `Generic`-like class -- otherwise I'm stuck at placing `Generic` somewhere in my inheritance chain.

**Reply by erictraut** [reply]
There are many expressions that can be evaluated at runtime that are not valid as type annotations. `uint8[3]` is an example. Pyright should emit an error here. This is a false negative, and it will be addressed in the next release.

If you want to use type annotations for things other than type expressions, you have a couple of options:
1. Use `Annotated` (introduced in PEP 593)[https://peps.python.org/pep-0593/]. This allows you to add arbitrary metadata to an annotation without interfering with static type analysis.
2. Propose an extension to the existing static type system in the [python/typing discussion forum](https://github.com/python/typing/discussions).

**Reply by ooliver1** [reply]
I previously used this in https://github.com/nextcord/nextcord/pull/980 since it worked just fine and all the issues seemed to point towards using this. I am aware of `typing.Annotated` but that is a lengthy `Annotated[int, Range(0.1, 10)]` over `Range[0.1, 10]`.
